### ISSUE TITLE
A Thorough Introduction to Distributed Systems  freeCodeCamp.org  Medium

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#4563]
A Thorough Introduction to Distributed Systems – freeCodeCamp.org – Medium<br>
<br>
http://bit.ly/2X5C7xf<br>
<br>
Stanislav Kozlovski<br>
<br>
<div><div data-tracking-context="postPage" data-post-id="3b91562c9b3c" data-source="post_page" data-collection-id="336d898217ee" data-scroll="native">
<section name="7971"><div>
<div>
<h2 name="d7d7">What is a Distributed System and why is it so complicated?</h2>
<div><div><a data-action-value="c520713b52a4" data-user-id="c520713b52a4" href="https://medium.com/@stanislavkozlovski?source=post_header_lockup" data-action="show-user-card" data-action-type="hover" dir="auto" data-action-source="post_header_lockup"><img alt="Go to the profile of Stanislav Kozlovski" src="https://cdn-images-1.medium.com/fit/c/100/100/1*aTbGPOSdDyke_7oOmfgpaQ.png"></a></div></div>
</div>
<div><figure name="9416" data-scroll="native"><div></div>
<figcaption>A bear contemplating distributed systems</figcaption></figure></div>
<div>
<h3 name="1b81">Table of Contents</h3>
<h4 name="c8d1">Introduction</h4>
<ol>
<li name="6ae1"><a href="https://medium.com/p/3b91562c9b3c#cef5" data-href="https://medium.com/p/3b91562c9b3c#cef5"><strong>What is a distributed system?</strong></a></li>
<li name="45dc"><a href="https://medium.com/p/3b91562c9b3c#97ac" data-href="https://medium.com/p/3b91562c9b3c#97ac"><strong>Why distribute a system?</strong></a></li>
<li name="72cd"><a href="https://medium.com/p/3b91562c9b3c#868b" data-href="https://medium.com/p/3b91562c9b3c#868b"><strong>Database scaling example</strong></a></li>
</ol>
<h4 name="3d7b"><a href="https://medium.com/p/3b91562c9b3c#8d9a" data-href="https://medium.com/p/3b91562c9b3c#8d9a">Distributed System Categories</a></h4>
<ol>
<li name="8690"><a href="https://medium.com/p/3b91562c9b3c#b201" data-href="https://medium.com/p/3b91562c9b3c#b201"><strong>Distributed Data Stores</strong></a></li>
<li name="af89"><a href="https://medium.com/p/3b91562c9b3c#b201" data-href="https://medium.com/p/3b91562c9b3c#b201"><strong>Distributed Computing</strong></a></li>
<li name="a17d"><a href="https://medium.com/p/3b91562c9b3c#dde2" data-href="https://medium.com/p/3b91562c9b3c#dde2"><strong>Distributed File Systems</strong></a></li>
<li name="a358"><a href="https://medium.com/p/3b91562c9b3c#8ecc" data-href="https://medium.com/p/3b91562c9b3c#8ecc"><strong>Distributed Messaging</strong></a></li>
<li name="3d1f"><a href="https://medium.com/p/3b91562c9b3c#1b6f" data-href="https://medium.com/p/3b91562c9b3c#1b6f"><strong>Distributed Applications</strong></a></li>
<li name="b0e0"><a href="https://medium.com/p/3b91562c9b3c#fbff" data-href="https://medium.com/p/3b91562c9b3c#fbff"><strong>Distributed Ledgers</strong></a></li>
</ol>
<h4 name="2878"><a href="https://medium.com/p/ad59df18f3c0#72cc" data-href="https://medium.com/p/ad59df18f3c0#72cc" target="_blank">Summary</a></h4>
<h3 name="7721">Introduction</h3>
<p name="77ec">With the ever-growing technological expansion of the world, distributed systems are becoming more and more widespread. They are a vast and complex field of study in computer science.</p>
<p name="8f03">This article aims to introduce you to distributed systems in a basic manner, showing you a glimpse of the different categories of such systems while not diving deep into the details.</p>
<h4 name="cef5">What is a distributed system?</h4>
<p name="1470">A distributed system in its most simplest definition is a group of computers working together as to appear as a single computer to the end-user.</p>
<p name="a552">These machines have a shared state, operate concurrently and can fail independently without affecting the whole system’s uptime.</p>
<p name="cd67">I propose we incrementally work through an example of distributing a system so that you can get a better sense of it all:</p>
<figure name="b6f7"><div></div>
<figcaption>A traditional stack</figcaption></figure><p name="2039">Let’s go with a database! Traditional databases are stored on the filesystem of one single machine, whenever you want to fetch/insert information in it — you talk to that machine directly.</p>
<p name="52db">For us to distribute this database system, we’d need to have this database run on multiple machines at the same time. The user must be able to talk to whichever machine he chooses and should not be able to tell that he is not talking to a single machine — if he inserts a record into node#1, node #3 must be able to return that record.</p>
<figure name="d698"><div></div>
<figcaption>An architecture that can be considered distributed</figcaption></figure><h4 name="97ac">Why distribute a system?</h4>
<p name="18b2">Systems are always distributed by necessity. The truth of the matter is — managing distributed systems is a complex topic chock-full of pitfalls and landmines. It is a headache to deploy, maintain and debug distributed systems, so why go there at all?</p>
<p name="6e27">What a distributed system enables you to do is <strong>scale horizontally</strong>. Going back to our previous example of the single database server, the only way to handle more traffic would be to upgrade the hardware the database is running on. This is called <strong>scaling vertically</strong>.</p>
<p name="7e0f">Scaling vertically is all well and good while you can, but after a certain point you will see that even the best hardware is not sufficient for enough traffic, not to mention impractical to host.</p>
<p name="98f0"><span name="anon_22a6fc3b66af" data-creator-ids="anon"><strong>Scaling horizontally</strong> simply means adding more computers rather than upgrading the hardware of a single one.</span></p>
<figure name="3b68"><div></div>
<figcaption>Horizontal scaling becomes <strong>much cheaper</strong> after a certain threshold</figcaption></figure><p name="83ee">It is significantly cheaper than vertical scaling after a certain threshold but that is not its main case for preference.</p>
<p name="d26a">Vertical scaling can only bump your performance up to the latest hardware’s capabilities. These capabilities prove to be <strong>insufficient</strong> for technological companies with moderate to big workloads.</p>
<p name="810f">The best thing about horizontal scaling is that you have no cap on how much you can scale — whenever performance degrades you simply add another machine, up to infinity potentially.</p>
<p name="3487">Easy scaling is not the only benefit you get from distributed systems. <strong><em>Fault tolerance</em></strong> and <strong><em>low latency</em></strong> are also equally as important.</p>
<p name="4272"><strong><em>Fault Tolerance </em></strong>— a cluster of ten machines across two data centers is inherently more fault-tolerant than a single machine. Even if one data center catches on fire, your application would still work.</p>
<p name="8194"><strong><em>Low Latency</em></strong> — The time for a network packet to travel the world is physically bounded by the speed of light. For example, the shortest possible time for a request‘s <strong>round-trip time</strong> (that is, go back and forth) in a fiber-optic cable between New York to Sydney is <a rel="nofollow noopener" href="https://www.oreilly.com/learning/primer-on-latency-and-bandwidth" data-href="https://www.oreilly.com/learning/primer-on-latency-and-bandwidth" target="_blank">160ms</a>. Distributed systems allow you to have a node in both cities, allowing traffic to hit the node that is closest to it.</p>
<p name="aa8d">For a distributed system to work, though, you need the software running on those machines to be specifically designed for running on multiple computers at the same time and handling the problems that come along with it. This turns out to be no easy feat.</p>
<h3 name="868b">Scaling our database</h3>
<p name="e70b">Imagine that our web application got insanely popular. Imagine also that our database started getting twice as much queries per second as it can handle. Your application would immediately start to decline in performance and this would get noticed by your users.</p>
<p name="8e0f">Let’s work together and make our database scale to meet our high demands.</p>
<p name="c02d">In a typical web application you normally read information much more frequently than you insert new information or modify old one.</p>
<figure name="30c8"><div></div></figure><p name="f749">There is a way to increase read performance and that is by the so-called <strong>Master-Slave Replication</strong> strategy. Here, you create two new database servers which sync up with the main one. The catch is that you can <strong>only read</strong> from these new instances.</p>
<p name="1dbc">Whenever you insert or modify information — you talk to the master database. It, in turn, asynchronously informs the slaves of the change and they save it as well.</p>
<p name="d40a">Congratulations, you can now execute 3x as much read queries! Isn’t this great?</p>
<h4 name="527f">Pitfall</h4>
<p name="02af"><strong>Gotcha!</strong> We immediately lost the <strong><em>C</em></strong> in our relational database’s <a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/ACID" data-href="https://en.wikipedia.org/wiki/ACID" target="_blank"><strong>ACID</strong></a> guarantees, which stands for Consistency.</p>
<p name="d7a5">You see, there now exists a possibility in which we insert a new record into the database, immediately afterwards issue a read query for it and get nothing back, as if it didn’t exist!</p>
<p name="1378">Propagating the new information from the master to the slave does not happen instantaneously. There actually exists a time window in which you can fetch stale information. If this were not the case, your write performance would suffer, as it would have to synchronously wait for the data to be propagated.</p>
<p name="0e66">Distributed systems come with a handful of trade-offs. This particular issue is one you will have to live with if you want to adequately scale.</p>
<p name="2b46"><strong>Continuing to Scale</strong></p>
<p name="fa6c">Using the slave database approach, we can horizontally scale our read traffic up to some extent. That’s great but we’ve hit a wall in regards to our write traffic — it’s still all in one server!</p>
<p name="02dc">We’re not left with much options here. We simply need to split our write traffic into multiple servers as one is not able to handle it.</p>
<p name="365c">One way is to go with a <a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Multi-master_replication" data-href="https://en.wikipedia.org/wiki/Multi-master_replication" target="_blank">multi-master replication strategy</a>. There, instead of slaves that you can only read from, you have multiple master nodes which support reads and writes. Unfortunately, this gets complicated real quick as you now have the ability to <a rel="noopener nofollow" href="http://datacharmer.blogspot.bg/2013/03/multi-master-data-conflicts-part-1.html" data-href="http://datacharmer.blogspot.bg/2013/03/multi-master-data-conflicts-part-1.html" target="_blank">create conflicts</a> (e.g insert two records with same ID).</p>
<p name="3780">Let’s go with another technique called <a href="https://medium.com/@jeeyoungk/how-sharding-works-b4dec46b3f6" data-href="https://medium.com/@jeeyoungk/how-sharding-works-b4dec46b3f6" target="_blank"><strong><em>sharding</em></strong></a><strong> </strong>(also called <strong><em>partitioning</em></strong>).</p>
<p name="1213">With sharding you split your server into multiple smaller servers, called <strong><em>shards. </em></strong>These shards all hold different records — you create a rule as to what kind of records go into which shard. It is very important to create the rule such that the data gets spread in an <strong>uniform way</strong>.</p>
<p name="7567">A possible approach to this is to define ranges according to some information about a record (e.g users with name A-D).</p>
<figure name="8154"><div></div></figure><p name="9207">This sharding key should be chosen very carefully, as the load is not always equal based on arbitrary columns. (e.g more people have a name starting with <em>C</em> rather than <em>Z</em>). A single shard that receives more requests than others is called a <strong><em>hot spot</em></strong> and must be avoided. Once split up, re-sharding data becomes incredibly expensive and can cause significant downtime, as was the case with <a rel="noopener nofollow" href="https://mashable.com/2010/10/05/foursquare-downtime-post-mortem/#qyp__Q9UDkqW" data-href="https://mashable.com/2010/10/05/foursquare-downtime-post-mortem/#qyp__Q9UDkqW" target="_blank">FourSquare’s infamous 11 hour outage</a>.</p>
<p name="eb47">To keep our example simple, assume our client (the Rails app) knows which database to use for each record. It is also worth noting that there are many strategies for sharding and this is a simple example to illustrate the concept.</p>
<p name="b8e5">We have won quite a lot right now — we can increase our write traffic <strong><em>N</em></strong> times where <strong><em>N</em></strong> is the number of shards. This practically gives us almost no limit — imagine how finely-grained we can get with this partitioning.</p>
<h4 name="8986">Pitfall</h4>
<p name="8ec7">Everything in Software Engineering is more or less a trade-off and this is no exception. Sharding is no simple feat and is best avoided <a rel="noopener nofollow" href="https://www.percona.com/blog/2009/08/06/why-you-dont-want-to-shard/" data-href="https://www.percona.com/blog/2009/08/06/why-you-dont-want-to-shard/" target="_blank">until really needed</a>.</p>
<p name="3506">We have now made queries by keys<em> </em><strong>other than the partitioned key</strong> incredibly inefficient (they need to go through all of the shards). SQL <code>JOIN</code> queries are even worse and complex ones become practically unusable.</p>
<h4 name="ef01">Decentralized vs Distributed</h4>
<p name="b441">Before we go any further I’d like to make a distinction between the two terms.</p>
<p name="62a2">Even though the words sound similar and can be concluded to mean the same logically, their difference makes a significant technological and political impact.</p>
<p name="ea0f"><strong><em>Decentralized</em></strong> is still<strong> <em>distributed</em></strong> in the technical sense, but the whole decentralized systems is not owned by one actor. No one company can own a decentralized system, otherwise it wouldn’t be decentralized anymore.</p>
<p name="f504">This means that most systems we will go over today can be thought of as <strong><em>distributed centralized systems</em></strong> — and that is what they’re made to be.</p>
<p name="274f">If you think about it — it is harder to create a decentralized system because then you need to handle the case where some of the participants are malicious. This is not the case with normal distributed systems, as you know you own all the nodes.</p>
<p name="b905"><em>Note: This definition has been </em><a rel="nofollow noopener" href="https://github.com/WebOfTrustInfo/rebooting-the-web-of-trust/issues/50#issuecomment-154995201" data-href="https://github.com/WebOfTrustInfo/rebooting-the-web-of-trust/issues/50#issuecomment-154995201" target="_blank"><em>debated</em></a><em> a lot and can be confused with others (peer-to-peer, federated). </em><a rel="nofollow noopener" href="https://ethereum.stackexchange.com/a/7829" data-href="https://ethereum.stackexchange.com/a/7829" target="_blank"><em>In early literature, it’s been defined differently as well.</em></a><em> Regardless, what I gave you as a definition is what I feel is the most widely used now that blockchain and cryptocurrencies popularized the term.</em></p>
<h3 name="8d9a">Distributed System Categories</h3>
<p name="8b14">We are now going to go through a couple of distributed system categories and list their largest publicly-known production usage. Bear in mind that most such numbers shown are outdated and are most probably significantly bigger as of the time you are reading this.</p>
<h3 name="b201">Distributed Data Stores</h3>
<p name="d5dd">Distributed Data Stores are most widely used and recognized as Distributed Databases. Most distributed databases are <a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/NoSQL" data-href="https://en.wikipedia.org/wiki/NoSQL" target="_blank">NoSQL</a> non-relational databases, limited to key-value semantics. They provide incredible performance and scalability at the cost of consistency or availability.</p>
<blockquote name="79ae">
<strong>Known Scale</strong> — <a rel="noopener nofollow" href="http://cassandra.apache.org/" data-href="http://cassandra.apache.org/" target="_blank">Apple is known to use 75,000 Apache Cassandra nodes storing over 10 petabytes of data,</a> back in 2015</blockquote>
<p name="31d9">We cannot go into discussions of distributed data stores without first introducing the <strong>CAP Theorem.</strong></p>
<h4 name="6c44">CAP Theorem</h4>
<p name="740d"><a rel="noopener nofollow" href="https://mwhittaker.github.io/blog/an_illustrated_proof_of_the_cap_theorem/" data-href="https://mwhittaker.github.io/blog/an_illustrated_proof_of_the_cap_theorem/" target="_blank">Proven way back in 2002</a>, the CAP theorem states that a distributed data store cannot simultaneously be consistent, available and partition tolerant.</p>
<figure name="0788"><div></div>
<figcaption>Choose 2 out of 3 (But not Consistency and Availability)</figcaption></figure><p name="8af6">Some quick definitions:</p>
<ul>
<li name="a6f0">
<strong>Consistency </strong>— What you read and write sequentially is what is expected (remember the gotcha with the database replication a few paragraphs ago?)</li>
<li name="9721">
<strong>Availability</strong> — the whole system does not die — every non-failing node always returns a response.</li>
<li name="6f28">
<strong>Partition Tolerant</strong> — The system continues to function and uphold its consistency/availability guarantees in spite of <a rel="noopener nofollow" href="https://www.symantec.com/security_response/glossary/define.jsp?letter=n&amp;word=network-partition" data-href="https://www.symantec.com/security_response/glossary/define.jsp?letter=n&amp;word=network-partition" target="_blank">network partitions</a>
</li>
</ul>
<p name="5a8e">In reality, partition tolerance must be a given for any distributed data store. As mentioned in many places, <a rel="noopener nofollow" href="https://codahale.com/you-cant-sacrifice-partition-tolerance/" data-href="https://codahale.com/you-cant-sacrifice-partition-tolerance/" target="_blank">one of which this great article</a>, you cannot have consistency and availability without partition tolerance.</p>
<p name="283a">Think about it: if you have two nodes which accept information and their connection dies — how are they both going to be available and simultaneously provide you with consistency? They have no way of knowing what the other node is doing and as such have can either become offline <em>(unavailable)</em> or work with stale information <em>(inconsistent)</em>.</p>
</div>
<div>
<p name="30ef">In the end you’re left to choose if you want your system to be strongly consistent or highly available <strong><em>under a network partition</em></strong>.</p>
<p name="48f3">Practice shows that most applications value availability more. You do not necessarily always need strong consistency. Even then, that trade-off is not necessarily made because you need the 100% availability guarantee, but rather because network latency can be an issue when having to synchronize machines to achieve strong consistency. These and more factors make applications typically opt for solutions which offer high availability.</p>
<p name="b59a">Such databases settle with the weakest consistency model — <strong><em>eventual consistency</em></strong> <a rel="noopener nofollow" href="https://hackernoon.com/eventual-vs-strong-consistency-in-distributed-databases-282fdad37cf7" data-href="https://hackernoon.com/eventual-vs-strong-consistency-in-distributed-databases-282fdad37cf7" target="_blank"><em>(strong vs eventual consistency explanation)</em></a>. This model guarantees that if no new updates are made to a given item, <strong>eventually</strong> all accesses to that item will return the latest updated value.</p>
<p name="97c5">Those systems provide <strong>BASE</strong> properties (as opposed to traditional databases’ ACID)</p>
<ul>
<li name="2d1d">
<strong>B</strong>asically <strong>A</strong>vailable — The system always returns a response</li>
<li name="e33b">
<strong>S</strong>oft state — The system could change over time, even during times of no input (due to eventual consistency)</li>
<li name="fc36">
<strong>E</strong>ventual consistency — In the absence of input, the data will spread to every node sooner or later — thus becoming consistent</li>
</ul>
<p name="dc6f">Examples of such available distributed databases — <a rel="noopener nofollow" href="http://cassandra.apache.org/" data-href="http://cassandra.apache.org/" target="_blank">Cassandra</a>, <a rel="noopener nofollow" href="http://basho.com/products/riak-kv/" data-href="http://basho.com/products/riak-kv/" target="_blank">Riak</a>, <a rel="noopener nofollow" href="http://www.project-voldemort.com/voldemort/" data-href="http://www.project-voldemort.com/voldemort/" target="_blank">Voldemort</a></p>
<p name="20b6">Of course, there are other data stores which prefer stronger consistency — <a rel="noopener nofollow" href="https://hbase.apache.org/" data-href="https://hbase.apache.org/" target="_blank">HBase</a>, <a rel="noopener nofollow" href="https://www.couchbase.com/" data-href="https://www.couchbase.com/" target="_blank">Couchbase</a>, <a rel="noopener nofollow" href="https://redis.io/" data-href="https://redis.io/" target="_blank">Redis,</a> <a rel="noopener nofollow" href="https://zookeeper.apache.org/" data-href="https://zookeeper.apache.org/" target="_blank">Zookeeper</a></p>
<p name="08da">The CAP theorem is worthy of multiple articles on its own — some regarding how you can <a rel="noopener nofollow" href="http://www.goland.org/blockchain_and_cap/" data-href="http://www.goland.org/blockchain_and_cap/" target="_blank">tweak a system’s CAP properties depending on how the client behaves</a> and others on <a rel="noopener nofollow" href="https://martin.kleppmann.com/2015/05/11/please-stop-calling-databases-cp-or-ap.html" data-href="https://martin.kleppmann.com/2015/05/11/please-stop-calling-databases-cp-or-ap.html" target="_blank">how it is not understood properly</a>.</p>
<h4 name="6881">Cassandra</h4>
<p name="87d4">Cassandra, as mentioned above, is a distributed No-SQL database which prefers the AP properties out of the CAP, settling with eventual consistency. I must admit this may be a bit misleading, as Cassandra is highly configurable — you can make it provide strong consistency at the expense of availability as well, but that is not its common use case.</p>
<p name="c2a4">Cassandra uses <a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Consistent_hashing" data-href="https://en.wikipedia.org/wiki/Consistent_hashing" target="_blank">consistent hashing</a> to determine which nodes out of your cluster must manage the data you are passing in. You set a <strong>replication factor</strong>, which basically states to how many nodes you want to replicate your data.</p>
<figure name="5953"><div></div>
<figcaption>Sample write</figcaption></figure><p name="2866">When reading, you will read from those nodes only.</p>
<p name="1c22">Cassandra is massively scalable, providing absurdly high write throughput.</p>
<figure name="cce7"><div></div>
<figcaption>Possibly biased diagram, showing writes per second benchmarks. <a rel="noopener nofollow" href="https://academy.datastax.com/planet-cassandra/nosql-performance-benchmarks" data-href="https://academy.datastax.com/planet-cassandra/nosql-performance-benchmarks" target="_blank">Taken from here.</a></figcaption></figure><p name="8cf0">Even though this diagram might be biased and it looks like it compares Cassandra to databases set to provide strong consistency (otherwise I can’t see why MongoDB would drop performance when upgraded from 4 to 8 nodes), this should still show what a properly set up Cassandra cluster is capable of.</p>
<p name="5531">Regardless, in the distributed systems trade-off which enables horizontal scaling and incredibly high throughput, Cassandra does not provide some fundamental features of ACID databases — namely, transactions.</p>
<h4 name="d6d7">Consensus</h4>
<p name="3b10">Database transactions are tricky to implement in distributed systems as they require each node to agree on the right action to take (abort or commit). This is known as <strong><em>consensus </em></strong>and it is a fundamental problem in distributed systems.</p>
<p name="2e7b">Reaching the type of agreement needed for the “transaction commit” problem is straightforward if the participating processes and the network are completely reliable. However, real systems are subject to a number of possible faults, such as process crashes, network partitioning, and lost, distorted, or duplicated messages.</p>
<p name="6895">This poses an issue — it has been <a rel="noopener nofollow" href="http://the-paper-trail.org/blog/a-brief-tour-of-flp-impossibility/" data-href="http://the-paper-trail.org/blog/a-brief-tour-of-flp-impossibility/" target="_blank">proven impossible</a> to guarantee that a correct consensus is reached within a bounded time frame on a non-reliable network.</p>
<p name="5dd8">In practice, though, there are algorithms that reach consensus on a non-reliable network pretty quickly. Cassandra actually provides <a rel="noopener nofollow" href="https://www.beyondthelines.net/databases/cassandra-lightweight-transactions/" data-href="https://www.beyondthelines.net/databases/cassandra-lightweight-transactions/" target="_blank">lightweight transactions</a> through the use of the <a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Paxos_%28computer_science%29" data-href="https://en.wikipedia.org/wiki/Paxos_(computer_science)" target="_blank">Paxos</a> algorithm for distributed consensus.</p>
<h3 name="622f">Distributed Computing</h3>
<p name="6d1f">Distributed computing is the key to the influx of Big Data processing we’ve seen in recent years. It is the technique of splitting an enormous task (e.g aggregate 100 billion records), of which no single computer is capable of practically executing on its own, into many smaller tasks, each of which can fit into a single commodity machine. You split your huge task into many smaller ones, have them execute on many machines in parallel, aggregate the data appropriately and you have solved your initial problem. This approach again enables you to scale horizontally — when you have a bigger task, simply include more nodes in the calculation.</p>
<blockquote name="9bbb">Known Scale — <a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Folding@home" data-href="https://en.wikipedia.org/wiki/Folding@home" target="_blank">Folding@Home</a> had <a rel="noopener nofollow" href="https://www.webcitation.org/6CWRdkzP0?url=http://fah-web.stanford.edu/cgi-bin/main.py?qtype=osstats2" data-href="https://www.webcitation.org/6CWRdkzP0?url=http://fah-web.stanford.edu/cgi-bin/main.py?qtype=osstats2" target="_blank">160k active machines in 2012</a>
</blockquote>
<p name="191e">An early innovator in this space was Google, which by necessity of their large amounts of data had to invent a new paradigm for distributed computation — MapReduce. They published a <a rel="noopener nofollow" href="http://static.googleusercontent.com/media/research.google.com/en/us/archive/mapreduce-osdi04.pdf" data-href="http://static.googleusercontent.com/media/research.google.com/en/us/archive/mapreduce-osdi04.pdf" target="_blank">paper on it in 2004</a> and the open source community later created <a rel="noopener nofollow" href="https://mapr.com/products/apache-hadoop" data-href="https://mapr.com/products/apache-hadoop" target="_blank">Apache Hadoop</a> based on it.</p>
<h4 name="3836">MapReduce</h4>
<p name="ae74">MapReduce can be simply defined as two steps — <a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Map_%28higher-order_function%29" data-href="https://en.wikipedia.org/wiki/Map_(higher-order_function)" target="_blank">mapping</a> the data and <a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Fold_%28higher-order_function%29" data-href="https://en.wikipedia.org/wiki/Fold_(higher-order_function)" target="_blank">reducing</a> it to something meaningful.</p>
<p name="88a2">Let’s get at it with an example again:</p>
<p name="0f4c">Say we are Medium and we stored our enormous information in a secondary distributed database for warehousing purposes. We want to fetch data representing the number of claps issued each day throughout April 2017 (a year ago).</p>
<p name="0267">This example is kept as short, clear and simple as possible, but imagine we are working with loads of data (e.g analyzing billions of claps). We won’t be storing all of this information on one machine obviously and we won’t be analyzing all of this with one machine only. We also won’t be querying the production database but rather some “warehouse” database built specifically for low-priority offline jobs.</p>
</div>
<div>
<p name="27e6">Each Map job is a separate node transforming as much data as it can. Each job traverses all of the data in the given storage node and maps it to a simple tuple of the date and the number one. Then, three intermediary steps <em>(which nobody talks about)</em> are done — Shuffle, Sort and Partition. They basically further arrange the data and delete it to the appropriate reduce job. As we’re dealing with big data, we have each Reduce job separated to work on a single date only.</p>
<p name="ab44">This is a good paradigm and surprisingly enables you to do a lot with it — you can chain multiple MapReduce jobs for example.</p>
<h4 name="57de">Better Techniques</h4>
<p name="735c">MapReduce is somewhat legacy nowadays and brings some problems with it. Because it works in batches (jobs) a problem arises where if your job fails — you need to restart the whole thing. A 2-hour job failing can really slow down your whole data processing pipeline and you do not want that in the very least, especially in peak hours.</p>
<p name="b0a9">Another issue is the time you wait until you receive results. In real-time analytic systems (which all have big data and thus use distributed computing) it is important to have your latest crunched data be as fresh as possible and certainly not from a few hours ago.</p>
<p name="13c4">As such, other <a rel="noopener nofollow" href="https://www.talend.com/blog/2017/08/28/lambda-kappa-real-time-big-data-architectures/" data-href="https://www.talend.com/blog/2017/08/28/lambda-kappa-real-time-big-data-architectures/" target="_blank">architectures have emerged</a> that address these issues. Namely <a rel="noopener nofollow" href="http://lambda-architecture.net/" data-href="http://lambda-architecture.net/" target="_blank">Lambda Architecture</a> (mix of batch processing and stream processing) and <a rel="noopener nofollow" href="http://milinda.pathirage.org/kappa-architecture.com/" data-href="http://milinda.pathirage.org/kappa-architecture.com/" target="_blank">Kappa Architecture</a> (only stream processing). These advances in the field have brought new tools enabling them — <a rel="noopener nofollow" href="https://kafka.apache.org/documentation/streams/" data-href="https://kafka.apache.org/documentation/streams/" target="_blank">Kafka Streams,</a> <a rel="noopener nofollow" href="https://spark.apache.org/" data-href="https://spark.apache.org/" target="_blank">Apache Spark</a>, <a rel="noopener nofollow" href="http://storm.apache.org/" data-href="http://storm.apache.org/" target="_blank">Apache Storm</a>, <a rel="noopener nofollow" href="http://samza.apache.org/" data-href="http://samza.apache.org/" target="_blank">Apache Samza</a>.</p>
<h3 name="dde2">Distributed File Systems</h3>
<p name="92ca">Distributed file systems can be thought of as distributed data stores. They’re the same thing as a concept — storing and accessing a large amount of data across a cluster of machines all appearing as one. They typically go hand in hand with Distributed Computing.</p>
<blockquote name="8c15">Known Scale — <a rel="noopener nofollow" href="https://www.slideshare.net/Hadoop_Summit/what-it-takes-to-run-hadoop-at-scale-yahoo-perspectives" data-href="https://www.slideshare.net/Hadoop_Summit/what-it-takes-to-run-hadoop-at-scale-yahoo-perspectives" target="_blank">Yahoo is known for running HDFS on over 42,000 nodes for storage of 600 Petabytes of data, way back in 2011</a>
</blockquote>
<p name="9df3">Wikipedia defines the difference being that distributed file systems allow files to be accessed using the same interfaces and semantics as local files, not through a custom API like the Cassandra Query Language <a rel="noopener nofollow" href="https://docs.datastax.com/en/cql/3.3/cql/cqlIntro.html" data-href="https://docs.datastax.com/en/cql/3.3/cql/cqlIntro.html" target="_blank">(CQL)</a>.</p>
<h4 name="2a03">HDFS</h4>
<p name="965d">Hadoop Distributed File System (HDFS) is the distributed file system used for distributed computing via the Hadoop framework. Boasting widespread adoption, it is used to store and replicate large files (GB or TB in size) across many machines.</p>
<p name="6860">Its architecture consists mainly of <strong><em>NameNodes</em></strong> and <strong><em>DataNodes</em></strong>. NameNodes are responsible for keeping metadata about the cluster, like which node contains which file blocks. They act as coordinators for the network by figuring out where best to store and replicate files, tracking the system’s health. DataNodes simply store files and execute commands like replicating a file, writing a new one and others.</p>
</div>
<div>
<p name="08a6">Unsurprisingly, HDFS is best used with Hadoop for computation as it provides data awareness to the computation jobs. Said jobs then get ran on the nodes storing the data. This leverages data locality — optimizes computations and reduces the amount of traffic over the network.</p>
<h4 name="09fb">IPFS</h4>
<p name="32b3"><a rel="noopener nofollow" href="https://ipfs.io/" data-href="https://ipfs.io/" target="_blank">Interplanetary File System (IPFS)</a> is an exciting new peer-to-peer protocol/network for a distributed file system. Leveraging <a href="https://medium.com/p/ad59df18f3c0#fbff" data-href="https://medium.com/p/ad59df18f3c0#fbff" target="_blank">Blockchain</a> technology, it boasts a completely decentralized architecture with no single owner nor point of failure.</p>
<p name="7dfc">IPFS offers a naming system (similar to DNS) called IPNS and lets users easily access information. It stores file via historic versioning, similar to how <a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Git" data-href="https://en.wikipedia.org/wiki/Git" target="_blank">Git</a> does. This allows for accessing all of a file’s previous states.</p>
<p name="5f95">It is still undergoing heavy development (v0.4 as of time of writing) but has already seen projects interested in building over it (<a rel="noopener nofollow" href="https://filecoin.io/" data-href="https://filecoin.io/" target="_blank">FileCoin</a>).</p>
<h3 name="8ecc">Distributed Messaging</h3>
<p name="b479">Messaging systems provide a central place for storage and propagation of messages/events inside your overall system. They allow you to decouple your application logic from directly talking with your other systems.</p>
<blockquote name="13d0">Known Scale — <a rel="noopener nofollow" href="https://engineering.linkedin.com/apache-kafka/how-we_re-improving-and-advancing-kafka-linkedin" data-href="https://engineering.linkedin.com/apache-kafka/how-we_re-improving-and-advancing-kafka-linkedin" target="_blank">LinkedIn’s Kafka cluster processed 1 trillion messages a day with peaks of 4.5 millions messages a second.</a>
</blockquote>
<figure name="7623"><div></div></figure><p name="8924">Simply put, a messaging platform works in the following way:</p>
<p name="c4b2">A message is broadcast from the application which potentially create it (called a <strong>producer</strong>), goes into the platform and is read by potentially multiple applications which are interested in it (called <strong>consumer</strong>s).</p>
<p name="b062">If you need to save a certain event to a few places (e.g user creation to database, warehouse, email sending service and whatever else you can come up with) a messaging platform is the cleanest way to spread that message.</p>
<p name="f34c">Consumers can either pull information out of the brokers (pull model) or have the brokers push information directly into the consumers (push model).</p>
<p name="dc67">There are a couple of popular top-notch messaging platforms:</p>
<p name="1a1c"><a rel="noopener nofollow" href="https://www.rabbitmq.com/" data-href="https://www.rabbitmq.com/" target="_blank">RabbitMQ</a> — Message broker which allows you finer-grained control of message trajectories via routing rules and other easily configurable settings. Can be called a smart broker, as it has a lot of logic in it and tightly keeps track of messages that pass through it. Provides settings for both <strong>AP</strong> and <strong>CP</strong> from <strong>CAP</strong>. Uses a push model for notifying the consumers.</p>
<p name="922b"><a rel="noopener nofollow" href="https://kafka.apache.org/" data-href="https://kafka.apache.org/" target="_blank">Kafka </a>— Message broker (and all out platform) which is a bit lower level, as in it does not keep track of which messages have been read and does not allow for complex routing logic. This helps it achieve amazing performance. In my opinion, this is the biggest prospect in this space with active development from the open-source community and support from the <a rel="noopener nofollow" href="https://www.confluent.io/blog" data-href="https://www.confluent.io/blog" target="_blank">Confluent team</a>. Kafka arguably has the most widespread use from top tech companies. <a rel="noopener nofollow" href="https://hackernoon.com/thorough-introduction-to-apache-kafka-6fbf2989bbc1" data-href="https://hackernoon.com/thorough-introduction-to-apache-kafka-6fbf2989bbc1" target="_blank">I wrote a thorough introduction to this, where I go into detail about all of its goodness.</a></p>
<p name="174b"><a rel="noopener nofollow" href="http://activemq.apache.org/" data-href="http://activemq.apache.org/" target="_blank">Apache ActiveMQ</a> — The oldest of the bunch, dating from 2004. Uses the JMS API, meaning it is geared towards Java EE applications. It got rewritten as <a rel="noopener nofollow" href="https://activemq.apache.org/artemis/" data-href="https://activemq.apache.org/artemis/" target="_blank">ActiveMQ Artemis</a>, which provides outstanding performance on par with Kafka.</p>
<p name="38ea"><a rel="noopener nofollow" href="https://aws.amazon.com/sqs/" data-href="https://aws.amazon.com/sqs/" target="_blank">Amazon SQS </a>— A messaging service provided by AWS. Lets you quickly integrate it with existing applications and eliminates the need to handle your own infrastructure, which might be a big benefit, as systems like Kafka are notoriously tricky to set up. Amazon also offers two similar services — <a rel="noopener nofollow" href="https://aws.amazon.com/sns/" data-href="https://aws.amazon.com/sns/" target="_blank">SNS</a> and <a rel="noopener nofollow" href="https://aws.amazon.com/amazon-mq/" data-href="https://aws.amazon.com/amazon-mq/" target="_blank">MQ</a>, the latter of which is basically ActiveMQ but managed by Amazon.</p>
<h3 name="1b6f">Distributed Applications</h3>
<p name="43dc">If you roll up 5 Rails servers behind a single load balancer all connected to one database, could you call that a distributed application? Recall my definition from up above:</p>
<blockquote name="ab0b">A distributed system is a group of computers working together as to appear as a single computer to the end-user. These machines have a shared state, operate concurrently and can fail independently without affecting the whole system’s uptime.</blockquote>
<p name="8790">If you count the database as a shared state, you could argue that this can be classified as a distributed system — but you’d be wrong, as you’ve missed the “<em>working together</em>” part of the definition.</p>
<p name="c600">A system is distributed only if the nodes communicate with each other to coordinate their actions.</p>
<p name="fea5">Therefore something like an application running its back-end code on a <a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Peer-to-peer" data-href="https://en.wikipedia.org/wiki/Peer-to-peer" target="_blank">peer-to-peer network</a> can better be classified as a distributed application. Regardless, this is all needless classification that serves no purpose but illustrate how fussy we are about grouping things together.</p>
<blockquote name="dd11">Known Scale — <a rel="noopener nofollow" href="https://torrentfreak.com/game-of-thrones-sets-new-torrent-swarm-record-140415/" data-href="https://torrentfreak.com/game-of-thrones-sets-new-torrent-swarm-record-140415/" target="_blank">BitTorrent swarm of 193,000 nodes for an episode of Game of Thrones, April, 2014</a>
</blockquote>
<h4 name="a8a6">Erlang Virtual Machine</h4>
<p name="5592">Erlang is a functional language that has great semantics for concurrency, distribution and fault-tolerance. The Erlang Virtual Machine itself handles the distribution of an Erlang application.</p>
<p name="0f24">Its model works by having many <strong>isolated</strong> <a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Light-weight_process" data-href="https://en.wikipedia.org/wiki/Light-weight_process" target="_blank">lightweight processes</a> all with the ability to talk to each other via a built-in system of message passing. This is called the <a rel="nofollow noopener" href="http://berb.github.io/diploma-thesis/original/054_actors.html" data-href="http://berb.github.io/diploma-thesis/original/054_actors.html" target="_blank"><strong><em>Actor Model</em></strong></a><strong><em> </em></strong>and the Erlang OTP libraries can be thought of as a distributed actor framework (along the lines of <a rel="nofollow noopener" href="https://akka.io/" data-href="https://akka.io/" target="_blank">Akka</a> for the JVM).</p>
<p name="f3dc">The model is what helps it achieve great concurrency rather simply — the processes are spread across the available cores of the system running them. Since this is indistinguishable from a network setting (apart from the ability to drop messages), Erlang’s VM can connect to other Erlang VMs running in the same data center or even in another continent. This swarm of virtual machines run one single application and handle machine failures via takeover (another node gets scheduled to run).</p>
<p name="b568">In fact, the distributed layer of the language was added in order to provide fault tolerance. Software running on a single machine is always at risk of having that single machine dying and taking your application offline. Software running on many nodes allows easier hardware failure handling, provided the application was built with that in mind.</p>
<h4 name="5139">BitTorrent</h4>
<p name="b8cb">BitTorrent is one of the most widely used protocol for transferring large files across the web via torrents. The main idea is to facilitate file transfer between different peers in the network without having to go through a main server.</p>
<p name="594f">Using a BitTorrent client, you connect to multiple computers across the world to download a file. When you open a .torrent file, you connect to a so-called <a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/BitTorrent_tracker" data-href="https://en.wikipedia.org/wiki/BitTorrent_tracker" target="_blank"><strong><em>tracker</em></strong></a>, which is a machine that acts as a coordinator. It helps with peer discovery, showing you the nodes in the network which have the file you want.</p>
<figure name="21f0"><div></div>
<figcaption>a sample network</figcaption></figure><p name="b7c3">You have the notions of two types of user, a <strong><em>leecher</em></strong> and a <strong><em>seeder</em></strong>. A leecher is the user who is downloading a file and a seeder is the user who is uploading said file.</p>
<p name="bd42">The funny thing about peer-to-peer networks is that you, as an ordinary user, have the ability to join and contribute to the network.</p>
<p name="6f32">BitTorrent and its precursors (<a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Gnutella" data-href="https://en.wikipedia.org/wiki/Gnutella" target="_blank">Gnutella</a>, <a rel="noopener nofollow" href="https://computer.howstuffworks.com/napster2.htm" data-href="https://computer.howstuffworks.com/napster2.htm" target="_blank">Napster</a>) allow you to voluntarily host files and upload to other users who want them. The reason BitTorrent is so popular is that it was the first of its kind to provide incentives for contributing to the network. <strong><em>Freeriding</em></strong>, where a user would only download files, was an issue with the previous file sharing protocols.</p>
<p name="89af">BitTorrent solved freeriding to an extent by making seeders upload more to those who provide the best download rates. It works by incentivizing you to upload while downloading a file. Unfortunately, after you’re done, nothing is making you stay active in the network. This causes a lack of seeders in the network who have the full file and as the protocol relies heavily on such users, solutions like <a rel="noopener nofollow" href="https://www.reddit.com/r/torrents/comments/2413bo/how_and_why_do_private_trackers_exist/ch2o1bo/" data-href="https://www.reddit.com/r/torrents/comments/2413bo/how_and_why_do_private_trackers_exist/ch2o1bo/" target="_blank">private trackers</a> came into fruition. Private trackers require you to be a member of a community (often invite-only) in order to participate in the distributed network.</p>
<p name="42cf">After advancements in the field, trackerless torrents were invented. This was an upgrade to the BitTorrent protocol that did not rely on centralized trackers for gathering metadata and finding peers but instead use new algorithms. One such instance is <a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Kademlia" data-href="https://en.wikipedia.org/wiki/Kademlia" target="_blank">Kademlia</a> (<a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Mainline_DHT" data-href="https://en.wikipedia.org/wiki/Mainline_DHT" target="_blank">Mainline DHT</a>), a distributed hash table (DHT) which allows you to find peers through other peers. In effect, each user performs a tracker’s duties.</p>
<h3 name="fbff">Distributed Ledgers</h3>
<p name="fca1">A distributed ledger can be thought of as an immutable, append-only database that is replicated, synchronized and shared across all nodes in the distributed network.</p>
<blockquote name="8a33">Known Scale — <a rel="noopener nofollow" href="https://etherscan.io/chart/tx" data-href="https://etherscan.io/chart/tx" target="_blank">Ethereum Network had a peak of 1.3 million transactions a day on January 4th, 2018</a>.</blockquote>
<p name="dc7e">They leverage the <a rel="noopener nofollow" href="https://martinfowler.com/eaaDev/EventSourcing.html" data-href="https://martinfowler.com/eaaDev/EventSourcing.html" target="_blank">Event Sourcing</a> pattern, allowing you to rebuild the ledger’s state at any time in its history.</p>
<h4 name="c274">Blockchain</h4>
<p name="f8f3">Blockchain is the current underlying technology used for distributed ledgers and in fact marked their start. This latest and greatest innovation in the distributed space enabled the creation of the first ever truly distributed payment protocol — Bitcoin.</p>
<p name="e4b5">Blockchain is a distributed ledger carrying an ordered list of all transactions that ever occurred in its network. Transactions are grouped and stored in blocks. The whole blockchain is essentially a <a rel="noopener nofollow" href="https://www.cs.cmu.edu/~adamchik/15-121/lectures/Linked%20Lists/linked%20lists.html" data-href="https://www.cs.cmu.edu/~adamchik/15-121/lectures/Linked%20Lists/linked%20lists.html" target="_blank">linked-list</a> of blocks <em>(hence the name)</em>. Said blocks are computationally expensive to create and are tightly linked to each other through cryptography.</p>
<p name="7c0c">Simply said, each block contains a special hash (that starts with X amount of zeroes) of the current block’s contents (in the form of a Merkle Tree) plus the previous block’s hash. This hash requires a lot of CPU power to be produced because the only way to come up with it is through brute-force.</p>
<figure name="74e2"><div></div>
<figcaption>Simplified blockchain</figcaption></figure><p name="fa4a"><strong>Miners</strong> are the nodes who try to compute the hash (via bruteforce). The miners all compete with each other for who can come up with a random string (called a <strong><em>nonce</em></strong>) which, when combine with the contents, produces the aforementioned hash. Once somebody finds the correct nonce — he broadcasts it to the whole network. Said string is then verified by each node on its own and accepted into their chain.</p>
<p name="97df">This translates into a system where it is absurdly costly to modify the blockchain and absurdly easy to verify that it is not tampered with.</p>
<p name="d637">It is costly to change a block’s contents because that would produce a different hash. Remember that each subsequent block‘s hash is dependent on it. If you were to change a transaction in the first block of the picture above — you would change the Merkle Root. This would in turn change the block’s hash (most likely without the needed leading zeroes) — that would change block #2’s hash and so on and so on. This means you’d need to brute-force a new nonce for every block after the one you just modified.</p>
<p name="a7c8">The network always trusts and replicates the longest valid chain. In order to cheat the system and <strong>eventually</strong> produce a longer chain you’d need more than 50% of the total CPU power used by all the nodes.</p>
<p name="dce4">Blockchain can be thought of as a distributed mechanism for <strong><em>emergent consensus</em></strong>. Consensus is not achieved explicitly — there is no election or fixed moment when consensus occurs. Instead, consensus is an <strong><em>emergent</em></strong> product of the asynchronous interaction of thousands of independent nodes, all following protocol rules.</p>
<p name="e340">This unprecedented innovation has recently become a boom in the tech space with people predicting it will mark the creation of the <a href="https://medium.com/@matteozago/why-the-web-3-0-matters-and-you-should-know-about-it-a5851d63c949" data-href="https://medium.com/@matteozago/why-the-web-3-0-matters-and-you-should-know-about-it-a5851d63c949" target="_blank">Web 3.0</a>. It is definitely the most exciting space in the software engineering world right now, filled with extremely challenging and interesting problems waiting to be solved.</p>
<h4 name="294f">Bitcoin</h4>
<p name="676e">What previous distributed payment protocols lacked was a way to practically prevent the <a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Double-spending" data-href="https://en.wikipedia.org/wiki/Double-spending" target="_blank">double-spending problem</a> in real time, in a distributed manner. Research has produced interesting propositions[1] but Bitcoin was the first to implement a practical solution with clear advantages over others.</p>
<p name="58c9">The double spending problem states that an actor (e.g Bob) cannot spend his single resource in two places. If Bob has $1, he should not be able to give it to both Alice and Zack — it is only one asset, it cannot be duplicated. It turns out it is really hard to truly achieve this guarantee in a distributed system. There are <a rel="noopener nofollow" href="https://arxiv.org/abs/0802.0832v1" data-href="https://arxiv.org/abs/0802.0832v1" target="_blank">some interesting mitigation approaches</a> predating blockchain, but they do not completely solve the problem in a practical way.</p>
<p name="85b2">Double-spending is solved easily by Bitcoin, as only one block is added to the chain at a time. Double-spending is impossible within a single block, therefore even if two blocks are created at the same time — only one will come to be on the eventual longest chain.</p>
<figure name="d636"><div></div></figure><p name="7e9b">Bitcoin relies on the difficulty of accumulating CPU power.</p>
<p name="d5cf">While in a voting system an attacker need only add nodes to the network (which is easy, as free access to the network is a design target), in a CPU power based scheme an attacker faces a physical limitation: getting access to more and more powerful hardware.</p>
<p name="7c5c">This is also the reason malicious groups of nodes need to control over 50% of the computational power of the network to actually carry any successful attack. Less than that, and the rest of the network will create a longer blockchain faster.</p>
<h4 name="73fa">Ethereum</h4>
<p name="7d8d">Ethereum can be thought of as a programmable blockchain-based software platform. It has its own cryptocurrency (Ether) which fuels the deployment of <strong><em>smart contracts </em></strong>on its blockchain.</p>
<p name="ab8a">Smart contracts are a piece of code stored as a single transaction in the Ethereum blockchain. To run the code, all you have to do is issue a transaction with a smart contract as its destination. This in turn makes the miner nodes execute the code and whatever changes it incurs. The code is executed inside the Ethereum Virtual Machine.</p>
<p name="4e8b"><strong><em>Solidity</em></strong>, Ethereum’s native programming language, is what’s used to write smart contracts. It is a turing-complete programming language which directly interfaces with the Ethereum blockchain, allowing you to query state like balances or other smart contract results. To prevent infinite loops, running the code requires some amount of Ether.</p>
<p name="0e74">As the blockchain can be interpreted as a series of <strong><em>state changes</em></strong>, a lot of Distributed Applications <a href="https://medium.com/the-mission/2018-the-year-of-dapps-dbe108860bcb" data-href="https://medium.com/the-mission/2018-the-year-of-dapps-dbe108860bcb" target="_blank">(DApps)</a> have been built on top of Ethereum and similar platforms.</p>
<h4 name="df49">Further usages of distributed ledgers</h4>
<p name="d709"><a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Proof_of_Existence" data-href="https://en.wikipedia.org/wiki/Proof_of_Existence" target="_blank"><strong><em>Proof of Existence</em></strong></a> — A service to anonymously and securely store proof that a certain digital document existed at some point of time. Useful for ensuring document integrity, ownership and timestamping.</p>
<p name="1883"><a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Decentralized_autonomous_organization" data-href="https://en.wikipedia.org/wiki/Decentralized_autonomous_organization" target="_blank"><strong><em>Decentralized Autonomous Organizations (DAO)</em></strong></a> — organizations which use blockchain as a means of reaching consensus on the organization’s improvement propositions. Examples are <a rel="noopener nofollow" href="https://www.dash.org/governance/" data-href="https://www.dash.org/governance/" target="_blank">Dash’s governance system</a>, <a rel="noopener nofollow" href="https://smartcash.cc/what-is-smartcash/" data-href="https://smartcash.cc/what-is-smartcash/" target="_blank">the SmartCash project</a></p>
<p name="a37a"><strong><em>Decentralized Authentication</em></strong> — Store your identity on the blockchain, enabling you to use <a rel="noopener nofollow" href="https://en.wikipedia.org/wiki/Single_sign-on" data-href="https://en.wikipedia.org/wiki/Single_sign-on" target="_blank">single sign-on</a> (SSO) everywhere. <a rel="noopener nofollow" href="https://sovrin.org/" data-href="https://sovrin.org/" target="_blank">Sovrin</a>, <a rel="noopener nofollow" href="https://www.civic.com/products/secure-identity-platform" data-href="https://www.civic.com/products/secure-identity-platform" target="_blank">Civic</a></p>
<p name="d587"><em>And many, many more. The distributed ledger technology really did open up endless possibilities. Some are most probably being invented as we speak!</em></p>
<h3 name="72cc">Summary</h3>
<p name="b597">In the short span of this article, we managed define what a distributed system is, why you’d use one and go over each category a little. Some important things to remember are:</p>
<ul>
<li name="134f">Distributed Systems are complex</li>
<li name="3851">They are chosen by necessity of scale and price</li>
<li name="d940">They are harder to work with</li>
<li name="0166">CAP Theorem — Consistency/Availability trade-off</li>
<li name="ed18">They have 6 categories — data stores, computing, file systems, messaging systems, ledgers, applications</li>
</ul>
<p name="694a">To be frank, we have barely touched the surface on distributed systems. I did not have the chance to thoroughly tackle and explain core problems like <a rel="noopener nofollow" href="https://www.cs.rutgers.edu/~pxk/417/notes/content/consensus.html" data-href="https://www.cs.rutgers.edu/~pxk/417/notes/content/consensus.html" target="_blank">consensus</a>, <a rel="noopener nofollow" href="http://www.cloudbus.org/papers/DataReplicationInDSChapter2006.pdf" data-href="http://www.cloudbus.org/papers/DataReplicationInDSChapter2006.pdf" target="_blank">replication strategies</a>, <a rel="noopener nofollow" href="https://swizec.com/blog/week-7-time-clocks-and-ordering-of-events-in-a-distributed-system/swizec/6444" data-href="https://swizec.com/blog/week-7-time-clocks-and-ordering-of-events-in-a-distributed-system/swizec/6444" target="_blank">event ordering &amp; time</a>, <a rel="noopener nofollow" href="http://blog.empathybox.com/post/19574936361/getting-real-about-distributed-system-reliability" data-href="http://blog.empathybox.com/post/19574936361/getting-real-about-distributed-system-reliability" target="_blank">failure tolerance</a>, <a rel="noopener nofollow" href="https://www.distributed-systems.net/my-data/papers/2007.osr.pdf" data-href="https://www.distributed-systems.net/my-data/papers/2007.osr.pdf" target="_blank">broadcasting a message across the network</a> and <a rel="noopener nofollow" href="http://the-paper-trail.org/blog/distributed-systems-theory-for-the-distributed-systems-engineer/" data-href="http://the-paper-trail.org/blog/distributed-systems-theory-for-the-distributed-systems-engineer/" target="_blank">others</a>.</p>
<h4 name="b559">Caution</h4>
<p name="b09d">Let me leave you with a parting forewarning:</p>
<p name="0845">You must stray away from distributed systems as much as you can. The complexity overhead they incur with themselves is not worth the effort if you can avoid the problem by either solving it in a different way or some other out-of-the-box solution.</p>
<figure name="7d72"><div></div></figure>
</div>
</div></section><section name="1b58"></section><section name="ab9a"><div><div>
<p name="5a4b">Thanks for taking the time to read through this long(~5600 words) article!</p>
<p name="5c85">If, by any chance, you found this informative or thought it provided you with value, please make sure to give it as many claps you believe it deserves and consider sharing with a friend who could use an introduction to this wonderful field of study.</p>
<p name="0b62"><em>~Stanislav Kozlovski</em></p>
</div></div></section><section name="89fe"><div><div>
<h4 name="c1fe">Update</h4>
<p name="3ddc">I currently work at <a rel="noopener nofollow" href="http://confluent.io" data-href="http://confluent.io" target="_blank">Confluent</a>. Confluent is a Big Data company founded by the creators of <a rel="noopener nofollow" href="https://hackernoon.com/thorough-introduction-to-apache-kafka-6fbf2989bbc1" data-href="https://hackernoon.com/thorough-introduction-to-apache-kafka-6fbf2989bbc1" target="_blank">Apache Kafka</a> themselves! I am immensely grateful for the opportunity they have given me — I currently work on Kafka itself, which is beyond awesome! We at Confluent help shape the whole open-source Kafka ecosystem, including a new managed Kafka-as-a-service cloud offering.</p>
<p name="6323">We are hiring for a lot of positions (especially SRE/Software Engineers) in Europe and the USA! If you are interested in working on Kafka itself, looking for new opportunities or just plain curious — make sure to message me on <a rel="noopener nofollow" href="https://twitter.com/StanKozlovski" data-href="https://twitter.com/StanKozlovski" target="_blank">Twitter</a> and I will share all the great perks that come from working in a bay area company.</p>
</div></div></section>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via Medium https://medium.com<br>
<br>
June 25, 2019 at 03:07PM
